File: scripts/import_leads.py
import os
import csv
import re
from datetime import datetime
try:
    import psycopg2
    from psycopg2.extras import execute_batch
except ImportError:
    print("Please install psycopg2-binary first: pip install psycopg2-binary")
    exit(1)

# Load database URL directly or from .env
# Using the direct URL for Supabase
DB_URL = "postgresql://postgres.tszrkdrzxozqyjvyiyec:Ved%40Upskilling@aws-1-ap-southeast-1.pooler.supabase.com:5432/postgres"

def main():
    csv_file = "blast - Sheet1.csv"
    
    if not os.path.exists(csv_file):
        print(f"Error: Could not find {csv_file} in this folder.")
        return

    print("Connecting to database...")
    try:
        conn = psycopg2.connect(DB_URL)
        cur = conn.cursor()
    except Exception as e:
        print("Failed to connect to database:", e)
        return

    # 1. Create the table exactly as Prisma expects it
    print("Ensuring CampaignLead table exists...")
    create_table_query = """
    CREATE TABLE IF NOT EXISTS "CampaignLead" (
        "id" SERIAL PRIMARY KEY,
        "phone" TEXT NOT NULL UNIQUE,
        "status" TEXT NOT NULL DEFAULT 'PENDING',
        "name" TEXT,
        "email" TEXT,
        "internshipInterest" TEXT,
        "lastContactedAt" TIMESTAMP(3) WITHOUT TIME ZONE,
        "createdAt" TIMESTAMP(3) WITHOUT TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP,
        "updatedAt" TIMESTAMP(3) WITHOUT TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP
    );
    """
    cur.execute(create_table_query)
    conn.commit()

    # 2. Read and parse CSV
    print("Reading CSV and parsing numbers...")
    valid_numbers = set()
    
    with open(csv_file, 'r', encoding='utf-8') as f:
        reader = csv.reader(f)
        for row in reader:
            for cell in row:
                # Keep only digits and plus signs
                clean_number = re.sub(r'[^0-9+]', '', str(cell).strip())
                if len(clean_number) >= 10:
                    valid_numbers.add(clean_number)

    if not valid_numbers:
        print("No valid phone numbers found in the CSV.")
        return

    print(f"Found {len(valid_numbers)} unique phone numbers to import.")

    # 3. Insert into Database
    insert_query = """
        INSERT INTO "CampaignLead" ("phone", "status", "updatedAt") 
        VALUES (%s, 'PENDING', %s) 
        ON CONFLICT ("phone") DO NOTHING;
    """
    
    now = datetime.utcnow()
    records = [(phone, now) for phone in valid_numbers]
    
    print("Uploading to Supabase... This might take a few seconds.")
    execute_batch(cur, insert_query, records, page_size=100)
    conn.commit()
    
    print(f"\nSuccess! All leads have been imported into the database.")
    
    cur.close()
    conn.close()

if __name__ == "__main__":
    main()

File: scripts/importLeads.ts
import { PrismaClient } from '@prisma/client';
import fs from 'fs';
import readline from 'readline';

const prisma = new PrismaClient();

async function importCSV() {
  const filePath = process.argv[2];
  
  if (!filePath) {
    console.error('Please provide a path to the CSV file as an argument.');
    console.error('Usage: npx ts-node scripts/importLeads.ts path/to/leads.csv');
    process.exit(1);
  }

  if (!fs.existsSync(filePath)) {
    console.error(`File not found: ${filePath}`);
    process.exit(1);
  }

  const fileStream = fs.createReadStream(filePath);
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  });

  let count = 0;
  let duplicates = 0;
  
  console.log('Starting import...');

  for await (const line of rl) {
    // The CSV has numbers in multiple columns. Split by comma.
    const columns = line.split(',');

    for (const col of columns) {
      const rawText = col.trim();
      // Only keep digits and plus signs
      const cleanNumber = rawText.replace(/[^0-9+]/g, '');

      // Check if it's a valid phone number (at least 10 digits)
      if (cleanNumber.length >= 10) {
        try {
          await prisma.campaignLead.create({
            data: {
              phone: cleanNumber,
              status: 'PENDING'
            }
          });
          count++;
          if (count % 100 === 0) {
            console.log(`Imported ${count} leads...`);
          }
        } catch (e: any) {
          // Prisma throws P2002 on unique constraint violation (duplicate phone)
          if (e.code === 'P2002') {
            duplicates++;
          } else {
            console.error(`Error inserting ${cleanNumber}:`, e.message);
          }
        }
      }
    }
  }

  console.log(`\nImport complete!`);
  console.log(`Successfully added: ${count} leads`);
  console.log(`Skipped (duplicates): ${duplicates} leads`);
  
  await prisma.$disconnect();
}

importCSV().catch(console.error);
